app/models/my_module.rb

```diff
@@ -375,6 +377,36 @@ def get_new_position
     { x: 0, y: positions.last[1] + HEIGHT }
   end
 
+  def completed?
+    state == 'completed'
+  end
+
+  # Mark task completed if all steps become completed
+  def check_completness
+    if protocol && protocol.steps.count > 0
+      completed = true
+      protocol.steps.find_each do |step|
+        completed = false unless step.completed
+      end
+      if completed
+        complete
+        save!
+        return true
+      end
+    end
+    false
+  end
+
+  def complete
+    self.state = 'completed'
```

Comment:
You can use here update_attributes ( state: 'xyz', completed_on: DateTime.now)
and avoid use save! in the check_completness method. Same goes for the uncomplete method.
Maybe rename complete to complete! and uncomplete!. The ! is used for methods that permanently changes its caller something in rails conventions